crates/telio-dns/src/nameserver.rs

```diff
@@ -113,26 +114,34 @@ impl LocalNameServer {
             let mut sending_buffer = sending_buffer.clone();
             let semaphore = semaphore.clone();
             tokio::spawn(async move {
-                match peer.decapsulate(
+                let res = peer.lock().await.decapsulate(
                     None,
                     receiving_buffer.get(..bytes_read).unwrap_or(&[]),
                     &mut sending_buffer,
-                ) {
+                );
+                match res {
                     // Handshake packets
                     TunnResult::WriteToNetwork(packet) => {
-                        if let Err(e) = socket.send_to(packet, dst_address).await {
+                        let res = socket.send_to(packet, dst_address).await;
+                        if let Err(e) = res {
                             telio_log_warn!("[DNS] Failed to send handshake packet  : {:?}", e);
                             return;
                         };
                         let mut temporary_buffer = [0u8; MAX_PACKET];
 
-                        while let TunnResult::WriteToNetwork(empty) =
-                            peer.decapsulate(None, &[], &mut temporary_buffer)
-                        {
+                        let mut last_res =
+                            peer.lock()
+                                .await
+                                .decapsulate(None, &[], &mut temporary_buffer);
+                        while let TunnResult::WriteToNetwork(empty) = last_res {
                             if let Err(e) = socket.send_to(empty, dst_address).await {
                                 telio_log_warn!("[DNS] Failed to send handshake packet  : {:?}", e);
                                 return;
                             };
+                            last_res =
```

Comment:
Why are you decapsulating here twice ?

Reply:
When it was done in the while condition the compiler complained about that - we cannot have lock locked while we await and `decapsulate` now needs to be called with lock, so I splitted it to two calls - I guess I could do that with `loop` and `break` if you would like that more, or maybe you have some idea of writing it in a better way?